### PR TITLE
[MLIR][TORCH] Add TorchToTosa lowering for aten.where.self op

### PR DESCRIPTION
**Pull request by AmosLewis**

### CODE REVIEW

**Comment by AmosLewis** [comment]
https://gist.github.com/AmosLewis/fa8662eef03c3379bda6a3974036abd1 for test

**Comment by AmosLewis** [comment]
In the test, it converts torch to tosa successfully. But why after that, it converts tosa to linalg? The tosa to linalg conversion arise the bug. @sjarus @eric-k256 
```
> (mlir_venv) nod% python -m e2e_testing.main -f 'ElementwiseAtenWhereSelf' --config=tosa -v 
> Compiling ElementwiseAtenWhereSelfModule_basic...
> FAIL - "ElementwiseAtenWhereSelfModule_basic"
> 
> Unexpected outcome summary:
> 
> ****** Failed tests - 1 tests
>     FAIL - "ElementwiseAtenWhereSelfModule_basic"
>         Compilation error: Traceback (most recent call last):
>           File "/home/chi/src/ubuntu20/shark/torch-mlir/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir_e2e_test/framework.py", line 290, in compile_and_run_test
>             compiled = config.compile(test.program_factory())
>           File "/home/chi/src/ubuntu20/shark/torch-mlir/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir_e2e_test/configs/tosa_backend.py", line 35, in compile
>             return self.backend.compile(module)
>           File "/home/chi/src/ubuntu20/shark/torch-mlir/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir_e2e_test/tosa_backends/linalg_on_tensors.py", line 57, in compile
>             run_pipeline_with_repro_report(
>           File "/home/chi/src/ubuntu20/shark/torch-mlir/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir/compiler_utils.py", line 73, in run_pipeline_with_repro_report
>             raise TorchMlirCompilerError(trimmed_message) from None
>         torch_mlir.compiler_utils.TorchMlirCompilerError: Lowering TOSA to Linalg-on-Tensors failed with the following diagnostics:
>         error: 'linalg.generic' op expected indexing_map #2 to have 4 dim(s) to match the number of loops
>         note: see current operation: 
>         %2 = "linalg.generic"(%1, %arg1, %arg2, %0) ({
>         ^bb0(%arg3: i1, %arg4: f32, %arg5: f32, %arg6: f32):
>         %3 = "arith.select"(%arg3, %arg4, %arg5) : (i1, f32, f32) -> f32
>         "linalg.yield"(%3) : (f32) -> ()
>         }) {indexing_maps = [affine_map<(d0, d1, d2, d3) -> (d0, d2, d3)>, affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>, affine_map<() -> ()>, affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>], iterator_types = ["parallel", "parallel", "parallel", "parallel"], operand_segment_sizes = array<i32: 3, 1>} : (tensor<1x5x5xi1>, tensor<1x12x5x5xf32>, tensor<f32>, tensor<1x12x5x5xf32>) -> tensor<1x12x5x5xf32>
```

**Comment by ramiro050** [comment]
I think the code that generates the indexing maps for the `linalg.generic` op is not handling zero-rank tensors correctly.

**Comment by AmosLewis** [comment]
> 

Yes. After replacing with rank, everything looks good.

**Comment by AmosLewis** [comment]
llvm to use [https://github.com/llvm/llvm-project/compare/main...AmosLewis:llvm-project:tosa-to-linalg](https://github.com/llvm/llvm-project/compare/main...AmosLewis:llvm-project:tosa-to-linalg)

**Review by sjarus — APPROVED** [review]

File: test/Conversion/TorchToTosa/basic.mlir [hunk 1]
```diff
@@ -913,3 +913,16 @@ func.func @torch.aten.to.dtype(%arg0: !torch.vtensor<[3,5],si64>) -> !torch.vten
...
+// CHECK-SAME:                                     %[[VAL_0:.*]]: tensor<1x1x5x5xi1>,
+// CHECK-SAME:                                     %[[VAL_1:.*]]: tensor<1x12x5x5xf32>,
+// CHECK-SAME:                                     %[[VAL_2:.*]]: tensor<f32>) -> tensor<1x12x5x5xf32> {
+// CHECK:           %[[VAL_3:.*]] = "tosa.select"(%[[VAL_0]], %[[VAL_1]], %[[VAL_2]]) : (tensor<1x1x5x5xi1>, tensor<1x12x5x5xf32>, tensor<f32>) -> tensor<1x12x5x5xf32>
+// CHECK:           return %[[VAL_3]] : tensor<1x12x5x5xf32>
+// CHECK:         }
+func.func @torch.aten.where.self(%arg0: !torch.vtensor<[1,1,5,5],i1>, %arg1: !torch.vtensor<[1,12,5,5],f32>, %arg2: !torch.vtensor<[],f32>) -> !torch.vtensor<[1,12,5,5],f32> {
+  %0 = torch.aten.where.self %arg0, %arg1, %arg2 : !torch.vtensor<[1,1,5,5],i1>, !torch.vtensor<[1,12,5,5],f32>, !torch.vtensor<[],f32> -> !torch.vtensor<[1,12,5,5],f32>
+  return %0 : !torch.vtensor<[1,12,5,5],f32>
+}
```
Comment (sjarus) [inline test/Conversion/TorchToTosa/basic.mlir:928]: This may be triggering your CI failure.

**Review by sjarus — CHANGES_REQUESTED** [review]

File: test/Conversion/TorchToTosa/basic.mlir (hunk 1 above)
Comment (AmosLewis) [reply test/Conversion/TorchToTosa/basic.mlir:928]: Need the llvm submodule update to this patch [https://github.com/llvm/llvm-project/commit/d325d2b407fcb7f37230b86fb291ef777497137d](https://github.com/llvm/llvm-project/commit/d325d2b407fcb7f37230b86fb291ef777497137d)

**Comment by AmosLewis** [comment]
Need this merged first.
[build: update llvm tag to 438e5918 #1475](https://github.com/llvm/torch-mlir/pull/1475)

**Comment by AmosLewis** [comment]
Finishe uplift the llvm to https://github.com/llvm/torch-mlir/pull/1502. But got a new build-test bug which is not from my patch.
```
  pip installing Pytorch..
  ERROR: Invalid requirement: '/main_checkout/torch-mlir/build_tools/../build_tools/python_deploy/wheelhouse/*'
  Hint: It looks like a path. File '/main_checkout/torch-mlir/build_tools/../build_tools/python_deploy/wheelhouse/*' does not exist.
  CMake Error at python/CMakeLists.txt:36 (message):
    Failed to run `build_libtorch.sh`
```

**Comment by ramiro050** [comment]
> Finish rebase and uplift the llvm to #1502. But got a new build-test out-of-tree bug which is not from my patch.
> 
> ```
>   pip installing Pytorch..
>   ERROR: Invalid requirement: '/main_checkout/torch-mlir/build_tools/../build_tools/python_deploy/wheelhouse/*'
>   Hint: It looks like a path. File '/main_checkout/torch-mlir/build_tools/../build_tools/python_deploy/wheelhouse/*' does not exist.
>   CMake Error at python/CMakeLists.txt:36 (message):
>     Failed to run `build_libtorch.sh`
> ```

Is this happening locally on your machine? The CI seems to be building fine.

**Comment by ramiro050** [comment]
> It was happening in CI (because of a bogus cache entry) so I deleted that cache entry and restarted the build, which then passed. All is fine now, but I'll keep an eye out for more failures.

Nice! Thanks for the quick fix!

**Comment by AmosLewis** [comment]
Here is the link to useful [torch_mlir_debug_command.txt](https://gist.github.com/AmosLewis/dd31ab37517977b1c499d06495b4adc2#file-torch_mlir_debug_command-txt) that will be helpful for boot camp and refer later in development of torch-mlir.

**Pull request merged**